service/src/main/java/bio/terra/drshub/services/DrsResolutionService.java

```diff
@@ -302,19 +307,51 @@ static String getObjectId(UriComponents uriComponents) {
         Optional.ofNullable(uriComponents.getPath()).orElse(""), StandardCharsets.UTF_8);
   }
 
-  private Optional<AccessMethod> getAccessMethod(DrsObject drsResponse, DrsProvider drsProvider) {
+  private Optional<AccessMethod> getAccessMethod(DrsObject drsResponse, DrsProvider drsProvider, ResolveFromEnum resolveFrom) {
+    Optional<AccessMethod> accessMethod = Optional.empty();
+    if (!isEmpty(drsResponse)) {
+      List<AccessMethod> accessMethods = getAccessMethods(drsResponse, drsProvider);
+      if (resolveFrom != null) {
+        accessMethod = getAccessMethodForCloud(accessMethods, resolveFrom);
+      }
+      // if there is no access method matching the resolveFrom cloud, or
+      // if the resolveFrom cloud was not specified, return a different access method
+      if (accessMethod.isEmpty() & !accessMethods.isEmpty()) {
+        accessMethod = Optional.of(accessMethods.get(0));
+      }
+    }
+    return accessMethod;
+  }
+
+  private Optional<AccessMethod> getAccessMethodForCloud(List<AccessMethod> accessMethods, ResolveFromEnum resolveFrom) {
+    Optional<AccessMethod> accessMethod;
+    if (resolveFrom.equals(ResolveFromEnum.AZURE)) {
+      accessMethod =
+          accessMethods.stream()
+              .filter(m -> m.getAccessId() != null)
+              .filter(m -> m.getAccessId().startsWith("az"))
+              .findFirst();
+    } else {
+      accessMethod =
+          accessMethods.stream()
+              .filter(m -> m.getType().toString().equals(resolveFrom.toString()))
+              .findFirst();
+    }
+    return accessMethod;
+  }
+
+  private List<AccessMethod> getAccessMethods(DrsObject drsResponse, DrsProvider drsProvider) {
+    ArrayList<AccessMethod> accessMethods = new ArrayList<>();
     if (!isEmpty(drsResponse)) {
       for (var methodConfig : drsProvider.getAccessMethodConfigs()) {
         var matchingMethod =
             drsResponse.getAccessMethods().stream()
                 .filter(m -> methodConfig.getType().getReturnedEquivalent() == m.getType())
-                .findFirst();
-        if (matchingMethod.isPresent()) {
-          return matchingMethod;
-        }
+                .toList();
+        accessMethods.addAll(matchingMethod);
       }
     }
-    return Optional.empty();
+    return accessMethods;
   }
```

Comment:
In general if you have `List<> l; x.stream().stream(add to l)` you can instead write it using `flatMap()`. In this case you can write:
```
  private List<AccessMethod> getAccessMethods(DrsObject drsResponse, DrsProvider drsProvider) {
    if (isEmpty(drsResponse)) {
      return List.of();
    }
    return drsProvider.getAccessMethodConfigs().stream()
        .flatMap(methodConfig -> drsResponse.getAccessMethods().stream()
            .filter(m -> methodConfig.getType().getReturnedEquivalent() == m.getType()))
        .toList();
  }
```